dali/python/nvidia/dali/plugin/base_iterator.py

```diff
@@ -382,6 +392,9 @@ def __next__(self):
         raise NotImplementedError
 
     def __iter__(self):
+        # avoid redundant reset when someone would call `iter()` on a new iterator
+        if self._counter != 0:
+            self.reset()
```

Comment:
The recommendation is to set size to `-1` when ES is used. Otherwise lot of strange things can happen (ES can raise StopIteration anytime and do it differently each epoch).
I don't think we can provide a consistent here no matter what.

Reply:
Sure, it makes sense. In that case maybe we should warn users against mixing size and ExternalSource (either with warning or in the docs/or both)?